c10/core/TensorImpl.cpp

```diff
@@ -37,14 +37,30 @@ TensorImpl::TensorImpl(TensorTypeId type_id, const caffe2::TypeMeta& data_type,
     : TensorImpl({}, type_id, data_type, is_variable) {
   // Variables, UndefinedTensors and SparseTensors don't have storages.
   if (!is_variable && type_id != UndefinedTensorId() && data_type.id() != caffe2::TypeIdentifier::uninitialized()
-      && type_id != SparseCPUTensorId() && type_id != SparseCUDATensorId()) {
+      && type_id != SparseCPUTensorId() && type_id != SparseCUDATensorId()
+      && type_id != MkldnnCPUTensorId()) {
     storage_ = Storage(data_type, 0, allocator, true);
   }
 }
 
 TensorImpl::TensorImpl(Storage&& storage, TensorTypeId type_id, bool is_variable)
     : TensorImpl(std::move(storage), type_id, storage.dtype(), is_variable) {}
 
+TensorImpl::TensorImpl(TensorTypeId type_id, const caffe2::TypeMeta& data_type, bool is_variable,
+                       c10::intrusive_ptr<c10::intrusive_ptr_target> opaque_handle, IntArrayRef sizes)
+  : TensorImpl({}, type_id, data_type, is_variable) {
+  AT_ASSERT(opaque_handle.get() != nullptr);
+  opaque_handle_ = opaque_handle;
+  auto dim = sizes.size();
+  sizes_.resize(dim);
+  for (size_t d = 0; d < dim; ++d) {
+    sizes_[d] = sizes[d];
+  }
+  is_contiguous_ = false;
+  allow_tensor_metadata_change_ = false;
+  refresh_numel();
+}
```

Comment:
we probably set strides_ to something meaningful too? or error out on access

Reply:
> we probably set strides_ to something meaningful too? or error out on access

@dzhulgakov Updated to disallow strides access. Thanks for the suggestion.

Reply:
> we probably set strides_ to something meaningful too? or error out on access

@dzhulgakov Thanks for the suggestion. Updated to disallow strides() access.